File: native/aloud-tts/ios/AloudProtocol.swift
import Foundation

// The Swift side of `contracts/commands.schema.json` — the exact shapes that
// cross the FFI boundary as JSON.
//
// Deliberately kept in its own file, importing nothing but Foundation, so the
// contract tests can compile it WITHOUT the Rust `AloudCoreFFI` module, the
// React module, an xcframework or a simulator. That is what lets
// `native/aloud-tts/ios/ContractTests` run these types against the shared
// `contracts/fixtures.json` as an ordinary `swift test` in CI — closing the
// gap called out in contracts/README.md, where Swift and Kotlin were the only
// bindings kept in step by review rather than by a test.
//
// The FFI wrapper that USES these types lives next door in AloudCore.swift.

/// Commands. Encodes to `{ "type": ..., ... }` to match the Rust `#[serde(tag)]`.
enum Command: Encodable {
    case play
    case pause
    case next
    case prev
    case getState
    case seekUnit(unit: Int)
    case seekByte(byte: Int)
    case wordBoundary(utf16Offset: Int)

    private enum CodingKeys: String, CodingKey {
        case type, unit, byte, utf16Offset
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .play: try c.encode("Play", forKey: .type)
        case .pause: try c.encode("Pause", forKey: .type)
        case .next: try c.encode("Next", forKey: .type)
        case .prev: try c.encode("Prev", forKey: .type)
        case .getState: try c.encode("GetState", forKey: .type)
        case let .seekUnit(unit):
            try c.encode("SeekUnit", forKey: .type)
            try c.encode(unit, forKey: .unit)
        case let .seekByte(byte):
            try c.encode("SeekByte", forKey: .type)
            try c.encode(byte, forKey: .byte)
        case let .wordBoundary(offset):
            try c.encode("WordBoundary", forKey: .type)
            try c.encode(offset, forKey: .utf16Offset)
        }
    }
}

struct Highlight: Codable, Equatable {
    let start: Int
    let end: Int
}

struct Snapshot: Codable, Equatable {
    let status: String
    let unit: Int
    let unitCount: Int
    let token: Int
    let tokenCount: Int
    let utterance: String
    let highlight: Highlight?

    /// Serialise back to a JSON dictionary for the RN bridge event payload.
    func asBridgePayload() -> [String: Any] {
        var dict: [String: Any] = [
            "status": status,
            "unit": unit,
            "unitCount": unitCount,
            "token": token,
            "tokenCount": tokenCount,
            "utterance": utterance,
            "highlight": NSNull(),
        ]
        if let h = highlight {
            dict["highlight"] = ["start": h.start, "end": h.end]
        }
        return dict
    }
}

struct CoreError: Codable, Error {
    let code: String
    let message: String
}

struct ErrorEnvelope: Codable {
    let error: CoreError
}

File: native/aloud-tts/android/src/main/java/com/aloud/tts/AloudProtocol.kt
package com.aloud.tts

import org.json.JSONObject

/**
 * The Kotlin side of `contracts/commands.schema.json` — the exact shapes that
 * cross the FFI boundary as JSON.
 *
 * Deliberately kept in its own file, depending on nothing but `org.json`, so the
 * contract tests can compile it WITHOUT JNA, the native `.so`, the Android SDK
 * or an emulator. That is what lets `android/contract-tests` run these types
 * against the shared `contracts/fixtures.json` as an ordinary JVM Gradle build
 * in CI — the Kotlin mirror of `ios/AloudProtocol.swift`.
 *
 * Note this costs no fidelity: Android's own unit tests run on the JVM against a
 * stub `android.jar` whose `org.json` throws unless the real artifact is added,
 * so the "proper" Android route would exercise the same `org.json` this does.
 *
 * The JNA wrapper that USES these types lives next door in AloudCore.kt.
 */

/** A parse error surfaced from the core as a typed exception. */
class CoreException(val code: String, override val message: String) : Exception(message)

data class Highlight(val start: Int, val end: Int)

data class Snapshot(
    val status: String,
    val unit: Int,
    val unitCount: Int,
    val token: Int,
    val tokenCount: Int,
    val utterance: String,
    val highlight: Highlight?,
)

/**
 * Command builders. The `type` strings here must match the Rust
 * `#[serde(tag = "type")]` variants exactly — a typo is rejected at runtime by
 * the core's deserialiser, never by the Kotlin compiler, which is precisely why
 * the contract test asserts them against the shared fixtures.
 */
object AloudCommand {
    fun play(): JSONObject = JSONObject().put("type", "Play")
    fun pause(): JSONObject = JSONObject().put("type", "Pause")
    fun next(): JSONObject = JSONObject().put("type", "Next")
    fun prev(): JSONObject = JSONObject().put("type", "Prev")
    fun getState(): JSONObject = JSONObject().put("type", "GetState")
    fun seekUnit(unit: Int): JSONObject = JSONObject().put("type", "SeekUnit").put("unit", unit)
    fun seekByte(byte: Int): JSONObject = JSONObject().put("type", "SeekByte").put("byte", byte)
    fun wordBoundary(utf16Offset: Int): JSONObject =
        JSONObject().put("type", "WordBoundary").put("utf16Offset", utf16Offset)
}

/**
 * Decode a core response object into a [Snapshot]. `internal` rather than
 * private so the contract tests, which compile this file directly, can exercise
 * it without going through JNA.
 */
internal fun JSONObject.toSnapshot(): Snapshot {
    val highlight = optJSONObject("highlight")?.let {
        Highlight(it.getInt("start"), it.getInt("end"))
    }
    return Snapshot(
        status = getString("status"),
        unit = getInt("unit"),
        unitCount = getInt("unitCount"),
        token = getInt("token"),
        tokenCount = getInt("tokenCount"),
        utterance = getString("utterance"),
        highlight = highlight,
    )
}
